src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -3543,6 +3543,10 @@ bool MarkAndCheckRequiresReflectionMethodBodyScanner (MethodBody body)
 			if (_compilerGeneratedMethodRequiresScanner.TryGetValue (body, out bool requiresReflectionMethodBodyScanner))
 				return requiresReflectionMethodBodyScanner;
 
+			// Make sure the method's body is processed, for compiler generated code we can get here before ProcessMethod is called
+			// and thus we need to make sure we operate on the optimized method body (to avoid marking code which is otherwise optimized away).
+			UnreachableBlocksOptimizer.ProcessMethod (body.Method);
```

Comment:
I thought about it, but it felt dangerous. I looked into this some more, and we have more bugs like this:

For example, `ProcessAnalysisAnnotationsForMethod` (which is called from `MarkMethod(MethodReference`, so way before any body processing is done on the method) calls `ShouldWarnForReflectionAccessToCompilerGeneratedCode` which does instructions scanning and may produce warnings if it sees references to annotated fields/methods.

If the method in question contains a feature switch condition, we won't avoid the warning, even thought we will eventually remove the code in question.

Basically this leads me to a bigger change:
* MarkStep should never access Method.Body directly - since we can't guarantee it's the right one.
* We will have to store some hashset of already processed bodies (basically what this change does), to avoid re-analyzing method bodies for const. prop and branch removal.
* Ever place where we access Body from the MarkStep needs to go through a helper which will trigger the const. prop and branch removal first.

The good news is that the UnreachableBlocksOptimizer is already written such that it can be called from pretty much anywhere and it will make sure to process things in the right order internally.